src/QsCompiler/CompilationManager/PerformanceTracking.cs

```diff
@@ -0,0 +1,202 @@
+        public enum Task
+        {
+            /// <summary>
+            /// Overall compilation process.
+            /// </summary>
+            OverallCompilation,
+
+            /// <summary>
+            /// Task that builds the compilation object.
+            /// </summary>
+            Build,
+
+            /// <summary>
+            /// Task that generates the compiled binary, DLL and docs.
+            /// </summary>
+            OutputGeneration,
+
+            /// <summary>
+            /// Task that loads references.
+            /// </summary>
+            ReferenceLoading,
+
+            /// <summary>
+            /// Task that performs rewrite steps.
+            /// </summary>
+            RewriteSteps,
+
+            /// <summary>
+            /// Task that loads sources.
+            /// </summary>
+            SourcesLoading,
+
+            /// <summary>
+            /// Task that replaces specific implementations as part of the 'Build' task.
+            /// </summary>
+            ReplaceTargetSpecificImplementations,
+
+            /// <summary>
+            /// Task that generates a binary as part of the 'OutputGeneration' task.
+            /// </summary>
+            BinaryGeneration,
+
+            /// <summary>
+            /// Task that generates a DLL as part of the 'OutputGeneration' task.
+            /// </summary>
+            DllGeneration,
+
+            /// <summary>
+            /// Task that generates documentation as part of the 'OutputGeneration' task.
+            /// </summary>
+            DocumentationGeneration,
+
+            /// <summary>
+            /// Task that serializes the syntax tree to be appended to the DLL as part of the 'OutputGeneration' task.
+            /// </summary>
+            SyntaxTreeSerialization,
+
+            /// <summary>
+            /// Task that loads data from references to a stream as part of the 'ReferenceLoading' task.
+            /// </summary>
+            LoadDataFromReferenceToStream,
+
+            /// <summary>
+            /// Task that initializes the deserializer object as part of the 'ReferenceLoading' task.
+            /// </summary>
+            DeserializerInit,
+
+            /// <summary>
+            /// Task that deserializes as part of the 'ReferenceLoading' task.
+            /// </summary>
+            SyntaxTreeDeserialization
+        }
+
+        /// <summary>
+        /// Used to raise a compilation task event.
+        /// </summary>
+        public static event CompilationTaskEventHandler CompilationTaskEvent;
+
+        /// <summary>
+        /// Whether a failure ocurred while tracking performance.
+        /// </summary>
+        public static bool FailureOccurred => FailureException != null;
+
+        /// <summary>
+        /// Exception that caused the failure to occur.
+        /// </summary>
+        #nullable enable
+        public static Exception? FailureException { get; private set; }
+        #nullable disable
```

Comment:
Would it be possible to use `#nullable enable` for the whole file?

Reply:
Sure :) making the change now.

Reply:
`#nullable enable` is now in effect for the whole file.

Reply:
Thanks!